File: py-usfm-parser/src/usfm_grammar/list_generator.py
'''Logics for USJ to list(table) conversions'''


class ListGenerator:
    '''Combines the methods used for List generattion from USJ'''
    def __init__(self):
        '''Variables shared by functions'''
        self.book = ""
        self.current_chapter = ""
        self.current_verse = ""
        self.list = [["Book","Chapter","Verse","Text","Type","Marker"]]

    def usj_to_list_id(self, obj):
        '''update book code'''
        self.book = obj['code']

    def usj_to_list_c(self, obj):
        '''Update current chapter'''
        self.current_chapter = obj['number']

    def usj_to_list_v(self, obj):
        '''Update current verse'''
        self.current_verse = obj['number']

    def usj_to_list(self, obj):
        '''Traverse the USJ dict and build the table in self.list'''
        if obj['type'] == "book":
            self.usj_to_list_id(obj)
        elif obj['type'] == "chapter":
            self.usj_to_list_c(obj)
        elif obj['type'] == "verse":
            self.usj_to_list_v(obj)
        marker_type = obj['type']
        marker_name = obj['marker'] if "marker" in obj else ''
        if marker_type == "USJ":
            # This would occur if the JSON got flatttened after removing paragraph markers
            marker_type = ""
        if 'content' in obj:
            for item in obj['content']:
                if isinstance(item, str):
                    self.list.append(
                        [self.book, self.current_chapter, self.current_verse,
                            item, marker_type, marker_name])
                else:
                    self.usj_to_list(item)

File: web-usfm-parser/src/listGenerator.js
class ListGenerator {
    /* Combines the methods used for List generation from USJ */
    constructor() {
        /* Variables shared by functions */
        this.book = "";
        this.currentChapter = "";
        this.currentVerse = "";
        this.list = [["Book", "Chapter", "Verse", "Text", "Type", "Marker"]];
    }

    usjToListId(obj) {
        /* Update book code */
        this.book = obj.code;
    }

    usjToListC(obj) {
        /* Update current chapter */
        this.currentChapter = obj.number;
    }

    usjToListV(obj) {
        /* Update current verse */
        this.currentVerse = obj.number;
    }

    usjToList(obj) {
        /* Traverse the USJ dict and build the table in this.list */
        if (obj.type === "book") {
            this.usjToListId(obj);
        } else if (obj.type === "chapter") {
            this.usjToListC(obj);
        } else if (obj.type === "verse") {
            this.usjToListV(obj);
        }

        let markerType = obj.type;
        let markerName = obj.marker ? obj.marker : '';
        
        if (markerType === "USJ") {
            // This would occur if the JSON got flattened after removing paragraph markers
            markerType = "";
        }

        if (obj.content) {
            for (let item of obj.content) {
                if (typeof item === "string") {
                    this.list.push([this.book, this.currentChapter, this.currentVerse, item, markerType, markerName]);
                } else {
                    this.usjToList(item);
                }
            }
        }
    }
}

export default ListGenerator;
